### PR TITLE
Use VecDeque for buffers instead of Vec.

### PR DESCRIPTION
**Pull request by canndrew**
Also add another benchmark.
This change seems to improve performance slightly (although less than I had expected) but definitely the more sensible way to implement it.

[Review on Reviewable](https://reviewable.io/reviews/maidsafe/rust-utp/18)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
r? @vinipsmaker

(maidsafe_highfive has picked a reviewer for you, use r? to override)

File: benches/socket.rs [hunk 1]
```diff
@@ -106,3 +108,44 @@ fn bench_transfer_one_megabyte(b: &mut Bencher) {
...
+        let mut server = iotry!(UtpSocket::bind(server_addr));
+
+        thread::spawn(move || {
+            let mut written_total = 0;
+            let mut client = iotry!(UtpSocket::connect(server_addr));
+            while written_total < len {
+                let chunk_size = rand::random::<u16>() as usize + 1;
+                let chunk_size = min(chunk_size, len - written_total);
+                //println!("written_total == {}; chunk_size == {}", written_total, chunk_size);
+                let written = iotry!(client.send_to(&data[written_total..(written_total + chunk_size)]));
```
Comment (vinipsmaker) [inline benches/socket.rs:131]: I'm not sure it's wise to benchmark like this. Each run will have very different characteristics. Can you change code in a way that we can at least control the seed? This way we can run the benchmark against two different implementations to see which one perform better under the same seed.

Comment (vinipsmaker) [inline benches/socket.rs:131]: In an ideal benchmark, we'd even move the `rand` call out of the measured code (i.e. collect all random numbers we'll need out of the loop and then used them inside the loop that measures time).

Comment (canndrew) [inline benches/socket.rs:131]: > I'm not sure it's wise to benchmark like this. Each run will have very different characteristics. Can you change code in a way that we can at least control the seed? This way we can run the benchmark against two different implementations to see which one perform better under the same seed.

Everytime you run `cargo bench` the benchmark gets run several hundred times and the results get averaged out. Besides which, the chunks are small enough and there's enough of them that I don't think it wouldn't make much of a difference anyway.

> In an ideal benchmark, we'd even move the rand call out of the measured code (i.e. collect all random numbers we'll need out of the loop and then used them inside the loop that measures time).

A better option would be to switch to a faster psuedo-random number generator. I could try sticking in something like https://github.com/deestan/rust-well1024a and see if it has any effect on the benchmarks.

Comment (vinipsmaker) [inline benches/socket.rs:131]: > Everytime you run `cargo bench` the benchmark gets run several hundred times and the results get averaged out.

Fair enough then.

**Pull request merged**